server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java

```diff
@@ -0,0 +1,141 @@
+import org.sonar.db.qualityprofile.OrgActiveRuleDto;
+import org.sonar.db.qualityprofile.QProfileDto;
+import org.sonar.server.exceptions.BadRequestException;
+import org.sonar.server.qualityprofile.index.ActiveRuleIndexer;
+
+import static org.sonar.server.ws.WsUtils.checkRequest;
+
+public class QProfileTreeImpl implements QProfileTree {
+
+  private final DbClient db;
+  private final RuleActivator ruleActivator;
+  private final System2 system2;
+  private final ActiveRuleIndexer activeRuleIndexer;
+
+  public QProfileTreeImpl(DbClient db, RuleActivator ruleActivator, System2 system2, ActiveRuleIndexer activeRuleIndexer) {
+    this.db = db;
+    this.ruleActivator = ruleActivator;
+    this.system2 = system2;
+    this.activeRuleIndexer = activeRuleIndexer;
+  }
+
+  @Override
+  public List<ActiveRuleChange> removeParentAndCommit(DbSession dbSession, QProfileDto profile) {
+    List<ActiveRuleChange> changes = removeParent(dbSession, profile);
+    activeRuleIndexer.commitAndIndex(dbSession, changes);
+    return changes;
+  }
+
+  @Override
+  public List<ActiveRuleChange> setParentAndCommit(DbSession dbSession, QProfileDto profile, QProfileDto parentProfile) {
+    List<ActiveRuleChange> changes = setParent(dbSession, profile, parentProfile);
+    activeRuleIndexer.commitAndIndex(dbSession, changes);
+    return changes;
+  }
+
+  private List<ActiveRuleChange> setParent(DbSession dbSession, QProfileDto profile, QProfileDto parent) {
+    checkRequest(parent.getLanguage().equals(profile.getLanguage()), "Cannot set the profile '%s' as the parent of profile '%s' since their languages differ ('%s' != '%s')",
+      parent.getKee(), profile.getKee(), parent.getLanguage(), profile.getLanguage());
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
+    if (parent.getKee().equals(profile.getParentKee())) {
+      return changes;
+    }
+
+    checkRequest(!isDescendant(dbSession, profile, parent), "Descendant profile '%s' can not be selected as parent of '%s'", parent.getKee(), profile.getKee());
+    changes.addAll(removeParent(dbSession, profile));
+
+    // set new parent
+    profile.setParentKee(parent.getKee());
+    db.qualityProfileDao().update(dbSession, profile);
+
+    List<OrgActiveRuleDto> parentActiveRules = db.activeRuleDao().selectByProfile(dbSession, parent);
+    Collection<RuleKey> ruleKeys = parentActiveRules.stream().map(ActiveRuleDto::getRuleKey).collect(MoreCollectors.toArrayList());
+    RuleActivationContext context = ruleActivator.createContextForUserProfile(dbSession, profile, ruleKeys);
+
+    for (ActiveRuleDto parentActiveRule : parentActiveRules) {
+      try {
+        RuleActivation activation = RuleActivation.create(parentActiveRule.getRuleKey(), null, null);
+        changes.addAll(ruleActivator.activate(dbSession, activation, context));
+      } catch (BadRequestException e) {
+        // for example because rule status is REMOVED
+        // TODO return errors
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvPmmckxG3vIth73&open=AWE-qvPmmckxG3vIth73)

[ISSUE_KEY]: AWE-qvPmmckxG3vIth73